### PR TITLE
UCS/CALLBACKQ: Avoid deadlocks when using from signal handler.

### PR DESCRIPTION
**Pull request by yosefe**
Blocking signals globally creates a bunch of problems, so it's simpler to block signals only for specific async context.
Also, avoid calling malloc/free since it's not signal safe - can deadlock as well (for same reason)

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1189/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1795/ for details.

**Comment by yosefe** [comment]
@alex-mikheev @shamisp pls review

File: test/gtest/ucs/test_callbackq.cc [hunk 1]
```diff
@@ -28,7 +28,7 @@ class test_callbackq : public ucs::test {
 
     virtual void init() {
         ucs::test::init();
-        ucs_status_t status = ucs_callbackq_init(&m_cbq, 64);
+        ucs_status_t status = ucs_callbackq_init(&m_cbq, 64, NULL);
         ASSERT_UCS_OK(status);
```
Comment (alex-mikheev) [inline test/gtest/ucs/test_callbackq.cc]: Probably it is a good idea to also test with a non null async context

Comment (yosefe) [inline test/gtest/ucs/test_callbackq.cc]: indeed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1196/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1803/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1804/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1197/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1805/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1198/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1806/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1807/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1201/ for details (Mellanox internal link).

File: src/ucs/datastruct/callbackq.h [hunk 2]
```diff
@@ -68,11 +68,14 @@ struct ucs_callbackq {
  *
  * @param  [in] cbq      Callback queue to initialize.
  * @param  [in] size     Callback queue size.
+ * @param  [in] async    If != NULL, make calling add/remove from this async
+ *                       context deadlock-free.
```
Comment (alex-mikheev) [inline src/ucs/datastruct/callbackq.h]: probably it is a good idea to document why deadlock can happen.

**Comment by alex-mikheev** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1202/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1808/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1203/ for details (Mellanox internal link).
